lib/process-pool.js

```diff
@@ -0,0 +1,131 @@
+// import maxCPU from './max-cpu';
+import Worker from './worker';
+import './ensure-promise';
+/* eslint-disable */
+import regeneratorRuntime from 'regenerator-runtime';
+/* eslint-enable */
+
+// TODO: this file is just a brainstorm sketch, it actually needs to be
+// implemented.
+
+const maxCPU = 4;
+
+export default class Pool {
+  constructor(options) {
+    // this.j = options && options.j || maxCPU;
+    this.workers = [];
+    this.idle = [];
+    // this.inbox = [];
+  }
+
+  isIdlePoolEmpty() {
+    return this.idle.length === 0;
+  }
+
+  isWorkersPoolEmpty() {
+    return this.workers.length === 0;
+  }
+
+  isMaxWorkers() {
+    return this.workers.length === maxCPU;
+  }
+
+  popWorker() {
+    let firstIndex = 0;
+    let lastIndex = this.workers.length - 1;
+
+    let result = this.workers[lastIndex];
+
+    [
+      this.workers[firstIndex],
+      this.workers[lastIndex]
+    ] = [
+      this.workers[lastIndex],
+      this.workers[firstIndex]
+    ];
+
+    return result;
+  }
+
+  popIdleWorker() {
+    return new Promise((resolve, reject) => {
+      if (this.isWorkersPoolEmpty()) {
+        let worker = new Worker(`${__dirname}/runner.js`);
+        console.log('ara:schedule - first run new worker');
+        this.workers.push(worker);
+        this.idle.push(worker);
+        resolve(worker);
+      }
+
+      if (this.isIdlePoolEmpty()) {
+        // this means all workers are busy
+        console.log('ara:schedule - busy');
+        setTimeout(() => {
```

Comment:
rather then spinning on idle, it may make more sense to store an queue (just an array) of work, so that when workers become idle they can pop up that queue.